tasty/generate_shapes.py

```diff
@@ -0,0 +1,283 @@
+            nodeshape_bn = BNode()
+            g.add((prop_bn, SH.qualifiedValueShape, nodeshape_bn))
+            g.add((nodeshape_bn, RDF.type, SH.NodeShape))
+            g.add((nodeshape_bn, SH['class'], namespaced_type))
+            if optional:
+                # For optionals, we set the severity to warning
+                g.add((prop_bn, SH.severity, SH.Warning))
+
+
+def add_mixins(g: Graph, namespaced_shape: URIRef, shape: dict, ns: Namespace):
+    if shape.get('shape-mixins') is not None:
+        for each_mixin in shape['shape-mixins']:
+            ns_mixin = ns[each_mixin]
+            if ns_mixin not in g.subjects():
+                print(f"Mixin {ns_mixin} not in graph")
+            else:
+                g.add((namespaced_shape, SH.node, ns_mixin))
+
+
+def add_qvs_property(g: Graph, each_path: dict, parent_namespaced_shape: URIRef, namespaced_path: URIRef) -> BNode:
+    property_bn = BNode()
+    g.add((parent_namespaced_shape, SH.property, property_bn))
+    add_sh_path(g, property_bn, each_path, namespaced_path)
+    g.add((property_bn, SH.qualifiedMinCount, Literal(1)))
+    g.add((property_bn, SH.qualifiedMaxCount, Literal(1)))
+    g.add((property_bn, SH.qualifiedValueShapesDisjoint, Literal(True)))
+    return property_bn
+
+
+def add_min_count_for_required_nodes(g: Graph, each_path: dict, namespaced_shape: URIRef, namespaced_path: URIRef):
+    """
+    Count the required number of 'types' and 'shapes' and add a minCount property constraint. Considers
+     inverse paths as well. Something like:
+            sh:property [
+            sh:path [ sh:inversePath phIoT:equipRef ] ;
+            sh:minCount 3 ;
+        ] ;
+    :param g:
+    :param each_path:
+    :param namespaced_shape:
+    :param namespaced_path:
+    :return:
+    """
+    required_nodes = len(each_path.get("types", [])) + len(each_path.get('shapes', []))
+    if required_nodes > 0:
+        # we define a minCount for each given relationship type
+        min_count_bn = BNode()
+        g.add((namespaced_shape, SH.property, min_count_bn))
+        g.add((min_count_bn, SH.minCount, Literal(required_nodes)))
+        add_sh_path(g, min_count_bn, each_path, namespaced_path)
+
+
+def add_sh_path(g: Graph, property_bn: BNode, each_path: dict, namespaced_path):
+    """
+    Add a path constraint, either inverse or direct
+    :param g:
+    :param property_bn:
+    :param each_path:
+    :param namespaced_path:
+    :return:
+    """
+    # Here we translate the path into a BNode if necessary (i.e. for inverse pathing)
+    if each_path.get('path-type') is not None and each_path['path-type'] == 'inverse':
+        path_bn = BNode()
+        g.add((property_bn, SH.path, path_bn))
+        g.add((path_bn, SH.inversePath, namespaced_path))
+    elif each_path.get('path-type') is not None and each_path['path-type'] == 'direct':
+        g.add((property_bn, SH.path, namespaced_path))
+
+
+def add_things(g, ns_shape, shape, ontology):
```

Comment:
"things" is unclear unless this is a concept in the semantic web world lol

Reply:
Yeah, totally. will do this in separate stuff, tracking:
https://github.com/NREL/tasty/issues/8